flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/PrintUtils.java

```diff
@@ -66,70 +77,102 @@ public static void printAsTableauForm(
 			PrintWriter printWriter) {
-		printAsTableauForm(tableSchema, it, printWriter, MAX_COLUMN_WIDTH, NULL_COLUMN, false);
+		printAsTableauForm(tableSchema, it, printWriter, MAX_COLUMN_WIDTH, NULL_COLUMN, false, false);
 	}
 
 	/**
 	 * Displays the result in a tableau form.
 	 *
+	 * <p><b>NOTE:</b> please make sure the data to print is small enough to be stored in java heap memory
+	 * if the column width is derived from content (`deriveColumnWidthByType` is false).
+	 *
 	 * <p>For example: (printRowKind is true)
-	 * +-------------+-------------+---------+-------------+
+	 * <pre>
+	 * +----------+-------------+---------+-------------+
 	 * | row_kind | boolean_col | int_col | varchar_col |
-	 * +-------------+-------------+---------+-------------+
+	 * +----------+-------------+---------+-------------+
 	 * |       +I |        true |       1 |         abc |
 	 * |       -U |       false |       2 |         def |
 	 * |       +U |       false |       3 |         def |
 	 * |       -D |      (NULL) |  (NULL) |      (NULL) |
-	 * +-------------+-------------+---------+-------------+
+	 * +----------+-------------+---------+-------------+
 	 * 4 rows in result
+	 * </pre>
+	 *
+	 * @param tableSchema The schema of the data to print
+	 * @param it The iterator for the data to print
+	 * @param printWriter The writer to write to
+	 * @param maxColumnWidth The max width of a column
+	 * @param nullColumn The string representation of a null value
+	 * @param deriveColumnWidthByType A flag to indicate whether the column width
+	 *        is derived from type (true) or content (false).
+	 * @param printRowKind A flag to indicate whether print row kind info
 	 */
 	public static void printAsTableauForm(
 			TableSchema tableSchema,
 			Iterator<Row> it,
 			PrintWriter printWriter,
 			int maxColumnWidth,
 			String nullColumn,
+			boolean deriveColumnWidthByType,
 			boolean printRowKind) {
-		List<String[]> rows = new ArrayList<>();
-
-		// fill field names first
-		final List<TableColumn> columns;
+		final List<TableColumn> columns = tableSchema.getTableColumns();
+		String[] columnNames = columns.stream().map(TableColumn::getName).toArray(String[]::new);
 		if (printRowKind) {
-			columns = Stream.concat(
-					Stream.of(TableColumn.of("row_kind", DataTypes.STRING())),
-					tableSchema.getTableColumns().stream()
-			).collect(Collectors.toList());
-		} else {
-			columns = tableSchema.getTableColumns();
+			columnNames = Stream.concat(Stream.of(ROW_KIND_COLUMN), Arrays.stream(columnNames)).toArray(String[]::new);
 		}
 
-		rows.add(columns.stream().map(TableColumn::getName).toArray(String[]::new));
-		while (it.hasNext()) {
-			rows.add(rowToString(it.next(), nullColumn, printRowKind));
+		final int[] colWidths;
+		if (deriveColumnWidthByType) {
+			colWidths = columnWidthsByType(
+					columns,
+					maxColumnWidth,
+					nullColumn,
+					printRowKind ? ROW_KIND_COLUMN : null);
+		} else {
+			//
+			List<Row> rows = new ArrayList<>();
```

Comment:
you are using `final` above but here and below you are not using it anymore, a reader now thinks that the variable is modified but this is not the case. It would be great to use `final` consistently.